### PR TITLE
Port grouping redo/v24

### PR DESCRIPTION
**Pull request by inashivb**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) tickets:
- https://redmine.openinfosecfoundation.org/issues/6795
- https://redmine.openinfosecfoundation.org/issues/6792

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1678

Previous PR: https://github.com/OISF/suricata/pull/10543

Changes since v23:
- rebased on top of latest master

### CODE REVIEW

File: src/interval-tree.h [hunk 1]
```diff
@@ -0,0 +1,789 @@
+/*	$NetBSD: interval-tree.h,v 1.8 2004/03/28 19:38:30 provos Exp $	*/
```
Comment (victorjulien) [inline src/interval-tree.h:1]: I think it would be good to add some info here:
- license of the file
- where did you take it from? Seems to come from one of the BSD's?

**Review by victorjulien — CHANGES_REQUESTED** [review]
Small update to commit message requested. Ready to go after that.

File: src/util-port-interval-tree.c [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+    if (a->port2 > b->max) {
+        b->max = a->port2;
+    }
+    if (a->port >= b->port) {
+        SCReturnInt(1);
+    }
+    SCReturnInt(-1);
+}
+
+// cppcheck-suppress nullPointerRedundantCheck
```
Comment (victorjulien) [inline src/util-port-interval-tree.c:50]: can you put this suppression in its own commit and add this to the git message:

```
src/util-interval-tree.c:51:1: warning: Either the condition 'tmp4!=NULL' is redundant or there is possible null pointer dereference: tmp4. [nullPointerRedundantCheck]
IRB_GENERATE(PI, SCIntervalNode, irb, PICompareAndUpdate);
^
src/util-interval-tree.c:51:1: note: Assuming that condition 'tmp4!=NULL' is not redundant
IRB_GENERATE(PI, SCIntervalNode, irb, PICompareAndUpdate);
^
src/util-interval-tree.c:51:1: note: Null pointer dereference
IRB_GENERATE(PI, SCIntervalNode, irb, PICompareAndUpdate);
^
src/util-interval-tree.c:51:1: warning: Either the condition 'tmp5!=NULL' is redundant or there is possible null pointer dereference: tmp5. [nullPointerRedundantCheck]
IRB_GENERATE(PI, SCIntervalNode, irb, PICompareAndUpdate);
^
src/util-interval-tree.c:51:1: note: Assuming that condition 'tmp5!=NULL' is not redundant
IRB_GENERATE(PI, SCIntervalNode, irb, PICompareAndUpdate);
^
src/util-interval-tree.c:51:1: note: Null pointer dereference
IRB_GENERATE(PI, SCIntervalNode, irb, PICompareAndUpdate);
^
src/util-interval-tree.c:51:1: warning: Either the condition 'oleft!=NULL' is redundant or there is possible null pointer dereference: oleft. [nullPointerRedundantCheck]
IRB_GENERATE(PI, SCIntervalNode, irb, PICompareAndUpdate);
^
src/util-interval-tree.c:51:1: note: Assuming that condition 'oleft!=NULL' is not redundant
IRB_GENERATE(PI, SCIntervalNode, irb, PICompareAndUpdate);
^
src/util-interval-tree.c:51:1: note: Null pointer dereference
IRB_GENERATE(PI, SCIntervalNode, irb, PICompareAndUpdate);
^
src/util-interval-tree.c:51:1: warning: Either the condition 'oright!=NULL' is redundant or there is possible null pointer dereference: oright. [nullPointerRedundantCheck]
IRB_GENERATE(PI, SCIntervalNode, irb, PICompareAndUpdate);
^
src/util-interval-tree.c:51:1: note: Assuming that condition 'oright!=NULL' is not redundant
IRB_GENERATE(PI, SCIntervalNode, irb, PICompareAndUpdate);
^
src/util-interval-tree.c:51:1: note: Null pointer dereference
IRB_GENERATE(PI, SCIntervalNode, irb, PICompareAndUpdate);
^
src/util-interval-tree.c:51:1: warning: Either the condition 'left!=NULL' is redundant or there is possible null pointer dereference: left. [nullPointerRedundantCheck]
IRB_GENERATE(PI, SCIntervalNode, irb, PICompareAndUpdate);
^
src/util-interval-tree.c:51:1: note: Assuming that condition 'left!=NULL' is not redundant
IRB_GENERATE(PI, SCIntervalNode, irb, PICompareAndUpdate);
^
src/util-interval-tree.c:51:1: note: Null pointer dereference
IRB_GENERATE(PI, SCIntervalNode, irb, PICompareAndUpdate);
^
```

Comment (victorjulien) [reply src/util-port-interval-tree.c:50]: ah that is slightly modified, let me get a fresh output

Comment (victorjulien) [reply src/util-port-interval-tree.c:50]: Use this
```
src/util-port-interval-tree.c:50:1: warning: Either the condition 'tmp!=NULL' is redundant or there is possible null pointer dereference: tmp. [nullPointerRedundantCheck]
IRB_GENERATE(PI, SCPortIntervalNode, irb, SCPortIntervalCompareAndUpdate);
^
src/util-port-interval-tree.c:50:1: note: Assuming that condition 'tmp!=NULL' is not redundant
IRB_GENERATE(PI, SCPortIntervalNode, irb, SCPortIntervalCompareAndUpdate);
^
src/util-port-interval-tree.c:50:1: note: Null pointer dereference
IRB_GENERATE(PI, SCPortIntervalNode, irb, SCPortIntervalCompareAndUpdate);
^
src/util-port-interval-tree.c:50:1: warning: Either the condition 'oleft!=NULL' is redundant or there is possible null pointer dereference: oleft. [nullPointerRedundantCheck]
IRB_GENERATE(PI, SCPortIntervalNode, irb, SCPortIntervalCompareAndUpdate);
^
src/util-port-interval-tree.c:50:1: note: Assuming that condition 'oleft!=NULL' is not redundant
IRB_GENERATE(PI, SCPortIntervalNode, irb, SCPortIntervalCompareAndUpdate);
^
src/util-port-interval-tree.c:50:1: note: Null pointer dereference
IRB_GENERATE(PI, SCPortIntervalNode, irb, SCPortIntervalCompareAndUpdate);
^
src/util-port-interval-tree.c:50:1: warning: Either the condition 'oright!=NULL' is redundant or there is possible null pointer dereference: oright. [nullPointerRedundantCheck]
IRB_GENERATE(PI, SCPortIntervalNode, irb, SCPortIntervalCompareAndUpdate);
^
src/util-port-interval-tree.c:50:1: note: Assuming that condition 'oright!=NULL' is not redundant
IRB_GENERATE(PI, SCPortIntervalNode, irb, SCPortIntervalCompareAndUpdate);
^
src/util-port-interval-tree.c:50:1: note: Null pointer dereference
IRB_GENERATE(PI, SCPortIntervalNode, irb, SCPortIntervalCompareAndUpdate);
^
src/util-port-interval-tree.c:50:1: warning: Either the condition 'left!=NULL' is redundant or there is possible null pointer dereference: left. [nullPointerRedundantCheck]
IRB_GENERATE(PI, SCPortIntervalNode, irb, SCPortIntervalCompareAndUpdate);
^
src/util-port-interval-tree.c:50:1: note: Assuming that condition 'left!=NULL' is not redundant
IRB_GENERATE(PI, SCPortIntervalNode, irb, SCPortIntervalCompareAndUpdate);
^
src/util-port-interval-tree.c:50:1: note: Null pointer dereference
IRB_GENERATE(PI, SCPortIntervalNode, irb, SCPortIntervalCompareAndUpdate);
^
```

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW2_autofp_stats_chk | | |
| .flow.end.tcp_state.established | 8 | 9 | 112.5% | 

Pipeline 18924

File: src/interval-tree.h (hunk 1 above)
Comment (inashivb) [reply src/interval-tree.h:1]: Please note that:
- There is the 2-clause BSD license a bit down further.
- I copied the pre-existing `tree.h` from `master` in this commit.
- I modified it for interval tree ops in the later commits.
- Currently, the `tree.h` that exists in *BSD is very different from this one. They implement rank balanced trees instead of red black trees (which is currently used by our code and these interval trees). So, idk if it's right to say that it comes from BSDs.

So, questions for you:
1. What should I add as license if the license is written down below?
2. Should I add that the file came from older versions of BSD`s `tree.h` to both `tree.h` and `interval-tree.h`?

Comment (victorjulien) [reply src/interval-tree.h:1]: ah no you can mention it's a copy of the in-tree tree.h
